content/posts/api-deployment-strategies.md

```diff
@@ -0,0 +1,341 @@
+{
+  "uri": "/hello",
+  "vars": [["arg_version", "==", "v2"]],
+  "upstream": {
+    "type": "roundrobin",
+    "nodes": {
+      "V2:80": 1
+    }
+  }
+}'
+```
+
+### In Request Headers
+
+Similar to routing based on query parameters, the below configuration routes requests based on headers. While `arg_<argument>` checks for `<argument>` in the query parameters, `http_<header>` checks for the `<header>` in the request headers:
+
+```shell
+curl http://127.0.0.1:9180/apisix/admin/routes/1 -X PUT -d '
+{
+  "uri": "/hello",
+  "vars": [["http_version", "==", "v1"]],
+  "upstream": {
+    "type": "roundrobin",
+    "nodes": {
+      "v1:80": 1
+    }
+  }
+}'
+```
+
+```shell
+curl http://127.0.0.1:9180/apisix/admin/routes/2 -X PUT -d '
+{
+  "uri": "/hello",
+  "vars": [["http_version", "==", "v2"]],
+  "upstream": {
+    "type": "roundrobin",
+    "nodes": {
+      "v2:80": 1
+    }
+  }
+}'
+```
+
+## Backwards Compatible APIs
+
+Multiple API versions cannot always get you far and sometimes, change is inevitable. A strategy to make changes to your API without breaking your client applications is to ensure backwards compatibility. i.e., the interface between your clients and the API remains the same while your API changes.
+
+For example, if you change your API endpoint (from `/old/api` to `/new/api`), an API gateway can redirect client requests to the new API endpoint without changes to the client applications.
+
+APISIX implements this through the [redirect](https://apisix.apache.org/docs/apisix/plugins/redirect/) and the [proxy-rewrite](https://apisix.apache.org/docs/apisix/plugins/proxy-rewrite/) plugins. The example below shows how you can configure redirects:
+
+```shell
+curl http://127.0.0.1:9180/apisix/admin/routes/1 -X PUT -d '
+{
+  "uri": "/old/api/hello",
+  "plugins": {
+    "redirect": {
+      "uri": "/new/api/hello",
+      "ret_code": 301
+    }
+  }
+}'
+```
+
+There could be more complex scenarios where the client-API interface changes. For instance, if the new API expects the key `fullName` instead of `firstName` and `lastName`, clients sending requests in the old API format would break. APISIX can handle such scenarios through the [body-transformer](https://apisix.apache.org/docs/apisix/plugins/body-transformer/) plugin.
+
+The example below shows how APISIX can transform the request body to make your APIs backwards compatible:
+
+```shell
+curl http://127.0.0.1:9180/apisix/admin/routes/1 -X PUT -d '
+{
+  "uri": "/hello",
+  "plugins": {
+    "body-transformer": {
+      "request": {
+        "template": "{ \"fullName\": \"{{ firstName }} {{ lastName }}\" }"
+      }
+    }
+  },
+  "upstream": {
+    "type": "roundrobin",
+    "nodes": {
+      "v2:80": 1
+    }
+  }
+}'
+```
+
+Now a request body like,
+
+```json
+{
+  "firstName": "Navendu",
+  "lastName": "Pottekkat"
+}
+```
+
+will be transformed to:
+
+```json
+{
+  "fullName": "Navendu Pottekkat"
+}
+```
+
```

Comment:
Great idea. However, you should also mention the cost in terms of resources and latency of such an approach.

It should be a temporary measure, not a solution